### ISSUE TITLE
[Feature Request] Generalize `beartype.vale.Is[...]` to accept `functools.partial(...)` objects

### ISSUE THREAD

**Issue by sylvorg** [beartype/beartype#363]
Hello, and sorry for not responding to your other hails!

Still have assignments to complete and finals to study for, but I'm submitting issues while taking breaks! 😹

Here's the latest one:

```python
from rich.traceback import install

install(show_locals=True)

from beartype.door import is_bearable
from beartype.vale import Is
from functools import partial
from typing import Annotated


def test(t):
    pass


is_bearable(object, Annotated[object, Is[partial(test)]])
```

```python
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /mnt/wsl/sylvorg/sylvorg/sylveon/siluam/oreo/./test16.py:18 in <module>                          │
│                                                                                                  │
│   15 def test(t): pass                                                                           │
│   16                                                                                             │
│   17                                                                                             │
│ ❱ 18 is_bearable(object, Annotated[object, Is[partial(test)]])                                   │
│   19                                                                                             │
│                                                                                                  │
│ ╭───────────────────────────── locals ──────────────────────────────╮                            │
│ │   Annotated = <class 'typing.Annotated'>                          │                            │
│ │     install = <function install at 0x7f92723fb600>                │                            │
│ │          Is = <beartype.vale._IsFactory object at 0x7f9271a676d0> │                            │
│ │ is_bearable = <function is_bearable at 0x7f9271a50720>            │                            │
│ │     partial = <class 'functools.partial'>                         │                            │
│ │        test = <function test at 0x7f9271f92f20>                   │                            │
│ ╰───────────────────────────────────────────────────────────────────╯                            │
│                                                                                                  │
│ /nix/store/ar1c9w057wxabggy5fynk4fmhs21nxz0-python3.11-beartype-0.16.4/lib/python3.11/site-packa │
│ ges/beartype/vale/_is/_valeis.py:332 in __getitem__                                              │
│                                                                                                  │
│   329 │   │   # Else, this class was subscripted by exactly one argument.                        │
│   330 │   │                                                                                      │
│   331 │   │   # If that callable is *NOT* a validator tester, raise an exception.                │
│ ❱ 332 │   │   die_unless_validator_tester(is_valid)                                              │
│   333 │   │   # Else, that callable is a validator tester.                                       │
│   334 │   │                                                                                      │
│   335 │   │   # Lambda function dynamically generating the machine-readable                      │
│                                                                                                  │
│ ╭──────────────────────────── locals ─────────────────────────────╮                              │
│ │ is_valid = functools.partial(<function test at 0x7f9271f92f20>) │                              │
│ │     self = <beartype.vale._IsFactory object at 0x7f9271a676d0>  │                              │
│ ╰─────────────────────────────────────────────────────────────────╯                              │
│                                                                                                  │
│ /nix/store/ar1c9w057wxabggy5fynk4fmhs21nxz0-python3.11-beartype-0.16.4/lib/python3.11/site-packa │
│ ges/beartype/vale/_util/_valeutilfunc.py:46 in die_unless_validator_tester                       │
│                                                                                                  │
│   43 │   # If this validator is either uncallable, a C-based callable, *OR* a                    │
│   44 │   # pure-Python callable accepting more or less than one parameter, raise                 │
│   45 │   # an exception.                                                                         │
│ ❱ 46 │   die_unless_func_args_len_flexible_equal(                                                │
│   47 │   │   func=validator_tester,                                                              │
│   48 │   │   func_args_len_flexible=1,                                                           │
│   49 │   │   exception_cls=BeartypeValeSubscriptionException,                                    │
│                                                                                                  │
│ ╭──────────────────────────────── locals ─────────────────────────────────╮                      │
│ │ validator_tester = functools.partial(<function test at 0x7f9271f92f20>) │                      │
│ ╰─────────────────────────────────────────────────────────────────────────╯                      │
│                                                                                                  │
│ /nix/store/ar1c9w057wxabggy5fynk4fmhs21nxz0-python3.11-beartype-0.16.4/lib/python3.11/site-packa │
│ ges/beartype/_util/func/arg/utilfuncargtest.py:90 in die_unless_func_args_len_flexible_equal     │
│                                                                                                  │
│    87 │   │   get_func_args_flexible_len)                                                        │
│    88 │                                                                                          │
│    89 │   # Number of flexible parameters accepted by this callable.                             │
│ ❱  90 │   func_args_len_flexible_actual = get_func_args_flexible_len(                            │
│    91 │   │   func=func,                                                                         │
│    92 │   │   is_unwrap=is_unwrap,                                                               │
│    93 │   │   exception_cls=exception_cls,                                                       │
│                                                                                                  │
│ ╭──────────────────────────────────────── locals ────────────────────────────────────────╮       │
│ │              exception_cls = <class 'beartype.roar.BeartypeValeSubscriptionException'> │       │
│ │                       func = functools.partial(<function test at 0x7f9271f92f20>)      │       │
│ │     func_args_len_flexible = 1                                                         │       │
│ │ get_func_args_flexible_len = <function get_func_args_flexible_len at 0x7f9271a50ea0>   │       │
│ │                  is_unwrap = True                                                      │       │
│ ╰────────────────────────────────────────────────────────────────────────────────────────╯       │
│                                                                                                  │
│ /nix/store/ar1c9w057wxabggy5fynk4fmhs21nxz0-python3.11-beartype-0.16.4/lib/python3.11/site-packa │
│ ges/beartype/_util/func/arg/utilfuncargget.py:123 in get_func_args_flexible_len                  │
│                                                                                                  │
│   120 │   '''                                                                                    │
│   121 │                                                                                          │
│   122 │   # Code object underlying the passed pure-Python callable unwrapped.                    │
│ ❱ 123 │   func_codeobj = get_func_codeobj(                                                       │
│   124 │   │   func=func,                                                                         │
│   125 │   │   is_unwrap=is_unwrap,                                                               │
│   126 │   │   exception_cls=exception_cls,                                                       │
│                                                                                                  │
│ ╭───────────────────────────────── locals ──────────────────────────────────╮                    │
│ │ exception_cls = <class 'beartype.roar.BeartypeValeSubscriptionException'> │                    │
│ │          func = functools.partial(<function test at 0x7f9271f92f20>)      │                    │
│ │     is_unwrap = True                                                      │                    │
│ ╰───────────────────────────────────────────────────────────────────────────╯                    │
│                                                                                                  │
│ /nix/store/ar1c9w057wxabggy5fynk4fmhs21nxz0-python3.11-beartype-0.16.4/lib/python3.11/site-packa │
│ ges/beartype/_util/func/utilfunccodeobj.py:128 in get_func_codeobj                               │
│                                                                                                  │
│   125 │   │   from beartype._util.func.utilfunctest import die_unless_func_python                │
│   126 │   │                                                                                      │
│   127 │   │   # Raise an exception.                                                              │
│ ❱ 128 │   │   die_unless_func_python(func=func, exception_cls=exception_cls)                     │
│   129 │   # Else, this callable is pure-Python and this code object exists.                      │
│   130 │                                                                                          │
│   131 │   # Return this code object.                                                             │
│                                                                                                  │
│ ╭────────────────────────────────────── locals ──────────────────────────────────────╮           │
│ │ die_unless_func_python = <function die_unless_func_python at 0x7f92719fa2a0>       │           │
│ │          exception_cls = <class 'beartype.roar.BeartypeValeSubscriptionException'> │           │
│ │                   func = functools.partial(<function test at 0x7f9271f92f20>)      │           │
│ │           func_codeobj = None                                                      │           │
│ │              is_unwrap = True                                                      │           │
│ ╰────────────────────────────────────────────────────────────────────────────────────╯           │
│                                                                                                  │
│ /nix/store/ar1c9w057wxabggy5fynk4fmhs21nxz0-python3.11-beartype-0.16.4/lib/python3.11/site-packa │
│ ges/beartype/_util/func/utilfunctest.py:110 in die_unless_func_python                            │
│                                                                                                  │
│   107 │   │   # Else, that callable is callable.                                                 │
│   108 │   │                                                                                      │
│   109 │   │   # Raise a human-readable exception.                                                │
│ ❱ 110 │   │   raise exception_cls(                                                               │
│   111 │   │   │   f'{exception_prefix}{repr(func)} not pure-Python function.')                   │
│   112 │   # Else, that callable is pure-Python.                                                  │
│   113                                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────── locals ───────────────────────────────────╮                 │
│ │    exception_cls = <class 'beartype.roar.BeartypeValeSubscriptionException'> │                 │
│ │ exception_prefix = ''                                                        │                 │
│ │             func = functools.partial(<function test at 0x7f9271f92f20>)      │                 │
│ ╰──────────────────────────────────────────────────────────────────────────────╯                 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
BeartypeValeSubscriptionException: Functools.partial(<function test at 0x7f9271f92f20>) not pure-Python function.
```

Again, sorry for the delay, but I gotta bounce! I'll answer everything as soon as possible! 😹

Thank you kindly for all the help, and for being a huge personal supporter! 😻

**Comment by leycec** [comment]
@sylvorg: the barely hanging-on student that just keeps on giving. But are they giving... *too much?* Only this woman-with-a-beard emoji knows. Clearly, she knows too much. :woman_beard: 

> Still have assignments to complete and finals to study for, but I'm submitting issues while taking breaks! 😹

You... you mad lad! If this is what you voluntarily do for fun, I can only imagine what your vacations look like. Please. Please don't break @beartype on your vacation. I'm crying exhaustedly here, bro. :weary: 

In other words, this feature request has been **resolved by 19c5fe844d840ed.** @beartype now accepts insane beartype validators like:
    
```python
from beartype.door import is_bearable
from beartype.typing import Annotated
from beartype.vale import Is
from functools import partial

def is_true(ignorable_arg, obj):
    '''
    Tester function returning :data:`True` only if the passed object
    evaluates to :data:`True` when coerced into a boolean and whose first
    parameter is ignorable.
    '''

    return bool(obj)

# Partial of the is_true() tester defined above, effectively ignoring the
# "ignorable_arg" parameter accepted by that tester.
is_true_partial = partial(is_true, 'Gods. This code is literally unreadable.')

# Beartype validator matching only objects that evaluate to "True".
Truthy = Annotated[object, Is[is_true_partial]]

assert is_bearable('', Truthy) is False
assert is_bearable('Even lies are true now, huh?', Truthy) is True
```

Is this valuable? I have no idea. Let's pretend I did something useful tonight so that I can sleep without self-recrimination.

**Comment by sylvorg** [comment]
Oh, I can assure you this, and beartype in general, has been _very_ valuable in my quest to build a better [`typer`](https://github.com/tiangolo/typer), [`feud`](https://github.com/eonu/feud), and [`cyclopts`](https://github.com/BrianPugh/cyclopts)! ... Oh, who am I kidding, I can't beat them... I've no idea what I'm doing half the time in python! 😭 I'll send y'all a sample when I'm done! 😹
